File: app/src/main/java/com/itachi1706/cheesecakeutilities/GeneralSettingsActivity.java
package com.itachi1706.cheesecakeutilities;


import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.Preference;

import androidx.appcompat.app.AppCompatActivity;

import com.itachi1706.appupdater.EasterEggResMusicPrefFragment;
import com.itachi1706.appupdater.SettingsInitializer;
import com.itachi1706.appupdater.Util.PrefHelper;
import com.itachi1706.cheesecakeutilities.Features.FingerprintAuth.AuthenticationActivity;
import com.itachi1706.cheesecakeutilities.Features.FingerprintAuth.BiometricCompatHelper;
import com.itachi1706.cheesecakeutilities.Features.UtilityManagement.ManageUtilityActivity;
import com.itachi1706.cheesecakeutilities.Util.CommonVariables;

import de.psdev.licensesdialog.LicensesDialog;


public class GeneralSettingsActivity extends AppCompatActivity {
    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);

        getFragmentManager().beginTransaction()
                .replace(android.R.id.content, new GeneralPreferenceFragment())
                .commit();
    }

    /**
     * This fragment shows general preferences only. It is used when the
     * activity is showing a two-pane settings UI.
     */
    public static class GeneralPreferenceFragment extends EasterEggResMusicPrefFragment {
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.pref_general);

            new SettingsInitializer().setFullscreen(true).explodeUpdaterSettings(getActivity(), R.drawable.notification_icon, CommonVariables.BASE_SERVER_URL,
                    getResources().getString(R.string.link_legacy), getResources().getString(R.string.link_updates), this);
            super.addEggMethods(true, preference -> {
                new LicensesDialog.Builder(getActivity()).setNotices(R.raw.notices)
                        .setIncludeOwnLicense(true).build().show();
                return true;
            });

            // Authentication processing
            final Preference fp_pw = findPreference("password_fp");

            sp = PrefHelper.getDefaultSharedPreferences(getActivity());
            updatePasswordViews(fp_pw);

            findPreference("testpw").setOnPreferenceClickListener(preference -> {
                startActivity(new Intent(getActivity(), AuthenticationActivity.class));
                return false;
            });

            findPreference("hide_util").setOnPreferenceClickListener(preference -> {
                startActivity(new Intent(getActivity(), ManageUtilityActivity.class));
                return false;
            });

            findPreference("util_settings").setOnPreferenceClickListener(preference -> {
                startActivity(new Intent(getActivity(), UtilitySettingsActivity.class));
                return false;
            });

            findPreference(BiometricCompatHelper.APP_BIOMETRIC_COMPAT_ENABLED).setOnPreferenceChangeListener((preference, newValue) -> {
                updatePasswordViews(fp_pw, (boolean) newValue, 0);
                return true;
            });

            findPreference(BiometricCompatHelper.SCREEN_LOCK_ENABLED).setOnPreferenceChangeListener((preference, newValue) -> {
                updatePasswordViews(fp_pw, (boolean) newValue, 1);
                return true;
            });

            findPreference("firebase_signin").setOnPreferenceClickListener(preference -> {
                Intent i = new Intent(getActivity(), FirebaseLoginActivity.class);
                i.putExtra("persist", true);
                startActivity(i);
                return false;
            });
        }

        @Override
        public void onResume() {
            super.onResume();

            updatePasswordViews(findPreference("password_fp"));
            boolean hasSL = BiometricCompatHelper.Companion.isScreenLockEnabled(getActivity());
            findPreference(BiometricCompatHelper.SCREEN_LOCK_ENABLED).setEnabled(hasSL);
            findPreference(BiometricCompatHelper.APP_BIOMETRIC_COMPAT_ENABLED).setEnabled(hasSL);
        }

        SharedPreferences sp;

        private void updatePasswordViews(Preference fp_pw) {
            updatePasswordViews(fp_pw, BiometricCompatHelper.Companion.requireFPAuth(sp), -1);
        }

        private void updatePasswordViews(Preference fp_pw, boolean val, int type) {
            boolean isScreenLock = BiometricCompatHelper.Companion.isScreenLockProtectionEnabled(getActivity()), isFP = BiometricCompatHelper.Companion.requireFPAuth(sp);
            switch (type) {
                case 0: isFP = val; break;
                case 1: isScreenLock = val; break;
            }
            String summary = "Unprotected";
            if (isScreenLock) {
                if (BiometricCompatHelper.Companion.isScreenLockEnabled(getActivity())) {
                    summary = "Protected with device screen lock";
                    if (isFP) {
                        if (BiometricCompatHelper.Companion.isBiometricFPRegistered(getActivity())) summary = "Protected with fingerprint + screen lock";
                        else summary += " (No fingerprint found on device)";
                    }
                }
                else summary = "Unprotected (No screen lock found)";
            } else if (isFP) {
                if (!BiometricCompatHelper.Companion.isScreenLockEnabled(getActivity())) summary = "Unprotected (No screen lock found)"; // No FP without a screen lock
                else if (BiometricCompatHelper.Companion.isBiometricFPRegistered(getActivity())) summary = "Protected with fingerprint";
                else summary = "Unprotected (No fingerprint found on device)";
            }

            fp_pw.setSummary(summary);
        }

        @Override
        public int getMusicResource() {
            return R.raw.hello;
        }

        @Override
        public String getStartEggMessage() {
            return "Hello!";
        }

        @Override
        public String getEndEggMessage() {
            return "Music Stopped";
        }

        @Override
        public String getStopEggButtonText() {
            return "SILENCE";
        }
    }
}

File: app/src/main/java/com/itachi1706/cheesecakeutilities/GeneralSettingsActivity.kt
package com.itachi1706.cheesecakeutilities

import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.preference.Preference
import com.itachi1706.appupdater.EasterEggResMultiMusicPrefFragment
import com.itachi1706.appupdater.SettingsInitializer
import com.itachi1706.cheesecakeutilities.features.biometricAuth.AuthenticationActivity
import com.itachi1706.cheesecakeutilities.features.biometricAuth.BiometricCompatHelper
import com.itachi1706.cheesecakeutilities.features.utilityManagement.ManageUtilityActivity
import com.itachi1706.cheesecakeutilities.util.CommonVariables
import com.itachi1706.helperlib.helpers.PrefHelper
import me.jfenn.attribouter.Attribouter

/**
 * Created by Kenneth on 24/7/2019.
 * for com.itachi1706.cheesecakeutilities in CheesecakeUtilities
 */
class GeneralSettingsActivity : AppCompatActivity() {
    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        supportFragmentManager.beginTransaction()
                .replace(android.R.id.content, GeneralPreferenceFragment())
                .commit()
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean { return if (item?.itemId == android.R.id.home) { finish(); true } else super.onOptionsItemSelected(item); }

    class GeneralPreferenceFragment : EasterEggResMultiMusicPrefFragment() {

        private lateinit var sp: SharedPreferences
        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            addPreferencesFromResource(R.xml.pref_general)

            SettingsInitializer().setFullscreen(true).explodeUpdaterSettings(activity, R.drawable.notification_icon, CommonVariables.BASE_SERVER_URL,
                    resources.getString(R.string.link_legacy), resources.getString(R.string.link_updates), this)
                    .setAboutApp(true) { Attribouter.from(context).show(); true }
                    .setIssueTracking(true, "https://itachi1706.atlassian.net/browse/CUTILAND")
                    .setBugReporting(true, "https://itachi1706.atlassian.net/servicedesk/customer/portal/3")
                    .setFDroidRepo(true, "fdroidrepos://fdroid.itachi1706.com/repo?fingerprint=B321F84BCAC7C296CF50923FF98965B11019BB5FD30C8B8F3A39F2F649AF9691")
                    .explodeInfoSettings(this)
            super.init()

            val bioPw: Preference? = findPreference("password_fp")
            sp = PrefHelper.getDefaultSharedPreferences(activity)
            updatePasswordViews(bioPw)

            findPreference<Preference>("testpw")?.setOnPreferenceClickListener {
                startActivity(Intent(activity, AuthenticationActivity::class.java))
                false
            }

            findPreference<Preference>("hide_util")?.setOnPreferenceClickListener {
                startActivity(Intent(activity, ManageUtilityActivity::class.java))
                false
            }

            findPreference<Preference>("util_settings")?.setOnPreferenceClickListener {
                startActivity(Intent(activity, UtilitySettingsActivity::class.java))
                false
            }

            findPreference<Preference>(BiometricCompatHelper.APP_BIOMETRIC_COMPAT_ENABLED)?.setOnPreferenceChangeListener { _, newValue ->
                updatePasswordViews(bioPw, newValue as Boolean, 0)
                true
            }

            findPreference<Preference>(BiometricCompatHelper.SCREEN_LOCK_ENABLED)?.setOnPreferenceChangeListener { _, newValue ->
                updatePasswordViews(bioPw, newValue as Boolean, 1)
                true
            }

            findPreference<Preference>("firebase_signin")?.setOnPreferenceClickListener {
                val i = Intent(activity, FirebaseLoginActivity::class.java)
                i.putExtra("persist", true)
                startActivity(i)
                false
            }

            findPreference<Preference>("app_theme")?.setOnPreferenceChangeListener { _, newValue ->
                updateDarkModeSetting(newValue.toString())
                true
            }

        }

        companion object {
            fun updateDarkModeSetting(newValue: String) {
                when (newValue) {
                    "light" -> PrefHelper.changeDarkModeTheme(AppCompatDelegate.MODE_NIGHT_NO, "Light")
                    "dark" -> PrefHelper.changeDarkModeTheme(AppCompatDelegate.MODE_NIGHT_YES, "Dark")
                    else -> {
                        // Set as battery saver default if P and below
                        PrefHelper.changeDarkModeTheme(if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY else AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM,
                                "default " + if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) "battery" else "system")
                    }
                }
            }
        }

        override fun onResume() {
            super.onResume()

            updatePasswordViews(findPreference("password_fp"))
            val hasSL = BiometricCompatHelper.isScreenLockEnabled(activity!!)
            findPreference<Preference>(BiometricCompatHelper.SCREEN_LOCK_ENABLED)?.isEnabled = hasSL
            findPreference<Preference>(BiometricCompatHelper.APP_BIOMETRIC_COMPAT_ENABLED)?.isEnabled = hasSL
        }

        private fun updatePasswordViews(bioPw: Preference?, value: Boolean = BiometricCompatHelper.requireBiometricAuth(sp), type: Int = -1) {
            var isScreenLock = BiometricCompatHelper.isScreenLockProtectionEnabled(activity!!)
            var isBiometric = BiometricCompatHelper.requireBiometricAuth(sp)

            when (type) {
                0 -> isBiometric = value
                1 -> isScreenLock = value
            }
            var summary = "Unprotected"
            if (isScreenLock) {
                if (BiometricCompatHelper.isScreenLockEnabled(activity!!)) {
                    summary = "Protected with device screen lock"
                    if (isBiometric) {
                        if (BiometricCompatHelper.isBiometricRegistered(activity!!)) summary = "Protected with biometrics + screen lock"
                        else summary += " (No biometric data found on device)"
                    }
                } else summary = "Unprotected (No screen lock found)"
            } else if (isBiometric) {
                summary = if (!BiometricCompatHelper.isScreenLockEnabled(activity!!)) "Unprotected (No screen lock found)" // No Biometrics without a screen lock
                else if (BiometricCompatHelper.isBiometricRegistered(activity!!)) "Protected with biometrics"
                else "Unprotected (No biometric data found on device)"
            }

            bioPw?.summary = summary
        }

        override fun getMusicResource(): Int {
            return R.raw.hello
        }

        override fun getStartEggMessage(): String {
            return "Hello!"
        }

        override fun getEndEggMessage(): String {
            return "Music Stopped"
        }

        override fun getStopEggButtonText(): String {
            return "SILENCE"
        }
    }
}